File: Test/FormMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include <memory>
#include <cstring>

#include <boost/regex.hpp>
#include <boost/utility.hpp>

#include "FormMain.h"

using std::unique_ptr;

using boost::wregex;
using boost::wcmatch;
using boost::regex_match;

using boost::noncopyable;

//#pragma alias "strrchr"    = "_strrchr"

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------

__fastcall TForm1::TForm1(TComponent* Owner)
    : TForm(Owner)
    , dmDatabase_( new TdmDatabase( 0 ) )
{

    ColorBox1->Selected = NDBGrid1->TitleFont->Color;
    ColorBox2->Selected = NDBGrid1->FixedColor;

    unique_ptr<TStringList> SLStyleNames( new TStringList() );
    for ( int Idx = 0 ; Idx < TStyleManager::StyleNames.Length ; ++Idx ) {
        SLStyleNames->Append( TStyleManager::StyleNames[Idx] );
    }
    comboboxStyle->Items->Assign( SLStyleNames.get() );
    comboboxStyle->ItemIndex =
        comboboxStyle->Items->IndexOf( StyleServices()->Name );

    dmDatabase_->FDQuery1->Open();
}
//---------------------------------------------------------------------------

void __fastcall TForm1::comboboxStyleChange(TObject *Sender)
{
    TStyleManager::SetStyle( comboboxStyle->Text );
}
//---------------------------------------------------------------------------

void __fastcall TForm1::NDBGrid1AfterAutoSet(TObject *Sender)
{
    TNDBGrid& Grid = dynamic_cast<TNDBGrid&>( *Sender );

    UnicodeString const OrderByClause = Grid.GetOrderByClause();
    TDataSource& DataSource = *Grid.DataSource;
    if ( TFDQuery* const DataSet = dynamic_cast<TFDQuery*>( DataSource.DataSet ) ) {
        unique_ptr<TStringList> const CmdText( new TStringList );
        CmdText->Assign( DataSet->SQL );

        class CloseOpenDS : public noncopyable {
        public:
            explicit CloseOpenDS( TFDQuery& DS ) : ds_( DS ) { ds_.Close(); }
            ~CloseOpenDS() throw() { ds_.Open(); }
        private:
            TFDQuery& ds_;
        } const volatile CloseOpenDSManager( *DataSet );

        ::OutputDebugString( Grid.OrderByPrefix.c_str() );

        wregex re( ( L"^\\s*" + Grid.OrderByPrefix + L".*$" ).c_str() );
        for ( int Idx = 0 ; Idx < CmdText->Count ; ++Idx ) {
            wcmatch what;

            if ( regex_match( CmdText->Strings[Idx].c_str(), what, re ) ) {
                CmdText->Strings[Idx] = OrderByClause;
                DataSet->SQL->Assign( CmdText.get() );

                return;
            }
        }
        CmdText->Append( OrderByClause );
        DataSet->SQL->Assign( CmdText.get() );
    }
}
//---------------------------------------------------------------------------


/*
void __fastcall TForm1::NDBGrid1DrawIndicator(TObject *Sender, int ACol, int ARow,
          const TRect &ARect, TGridDrawState AState, TNDBGridIndicatorState IndicatorState)

{
    NDBGrid1->DefaultDrawCell( ACol, ARow, ARect, AState );
    TRect XR( ARect.Left, ARect.Top, ARect.Left + 4, ARect.Top + 4 );
    NDBGrid1->Canvas->Brush->Color = clYellow;
    NDBGrid1->Canvas->FillRect( XR );
}
*/
//---------------------------------------------------------------------------

void __fastcall TForm1::Button1Click(TObject *Sender)
{
    NDBGrid1->TitleFont->Color = ColorBox1->Selected;
    DBGrid1->TitleFont->Color = ColorBox1->Selected;
}
//---------------------------------------------------------------------------

/*
void __fastcall TForm1::NDBGrid1DrawTitle(TObject *Sender, int ACol, int ARow, const TRect &ARect,
          TGridDrawState AState)
{
    if ( NDBGrid1->Options.Contains( dgIndicator ) ) {
        if ( ACol ) {
            NDBGrid1->DrawTitleCell(
                ACol, ARow, ARect, NDBGrid1->Columns->Items[ACol-1], AState
            );
        }
        else {
            NDBGrid1->DefaultDrawCell( ACol, ARow, ARect, AState );
        }
    }
    else
        NDBGrid1->DrawTitleCell(
            ACol, ARow, ARect, NDBGrid1->Columns->Items[ACol], AState
        );
    TRect XR( ARect.Left, ARect.Top, ARect.Left + 4, ARect.Top + 4 );
    NDBGrid1->Canvas->Brush->Color = AState.Contains( gdPressed ) ? clRed : clGreen;
    NDBGrid1->Canvas->FillRect( XR );
}
*/
//---------------------------------------------------------------------------

void __fastcall TForm1::Button2Click(TObject *Sender)
{
    NDBGrid1->FixedColor = ColorBox2->Selected;
    DBGrid1->FixedColor = ColorBox2->Selected;

}
//---------------------------------------------------------------------------

void __fastcall TForm1::NDBGrid1MouseMove(TObject *Sender, TShiftState Shift, int X,
          int Y)
{
    if ( !NDBGrid1->Sizing( X, Y ) && Shift.Contains( ssLeft ) ) {
        NDBGrid1->BeginDrag( false );
    }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::DBGrid1MouseMove(TObject *Sender, TShiftState Shift, int X,
          int Y)
{
//    if ( !DBGrid1->Sizing( X, Y ) && Shift.Contains( ssLeft ) ) {
//        DBGrid1->BeginDrag( false );
//    }
}
//---------------------------------------------------------------------------

File: Test/Delphi/FormMain.pas
unit FormMain;

interface

uses
  Winapi.Windows,
  System.SysUtils, System.Classes, System.Actions, System.RegularExpressions,
  Vcl.Controls, Vcl.Forms, Vcl.StdCtrls, Vcl.DBGrids, Vcl.Grids,
  Vcl.ActnList, Vcl.ActnMan, Vcl.ActnCtrls, Vcl.ActnMenus, Vcl.ActnPopup,
  Vcl.PlatformDefaultStyleActnCtrls, Vcl.StdActns,
  Vcl.Buttons, Vcl.ComCtrls, Vcl.DBCtrls, Vcl.ExtCtrls, Vcl.ImgList,
  Vcl.Menus, Vcl.Themes, Vcl.ToolWin,
  Data.DB,
  FireDAC.Comp.Client, FireDAC.Comp.UI,
  FireDAC.Stan.Intf, FireDAC.UI.Intf, FireDAC.VCLUI.Wait,
  NDBGrid,
  DataMod;

type
  TForm1 = class(TForm)
    DataSource1: TDataSource;
    DBGrid1: TDBGrid;
    Panel1: TPanel;
    DBNavigator1: TDBNavigator;
    Panel2: TPanel;
    NDBGrid1: TNDBGrid;
    Splitter1: TSplitter;
    Panel3: TPanel;
    ActionManager1: TActionManager;
    FileExit1: TFileExit;
    ActionMainMenuBar1: TActionMainMenuBar;
    PopupMenu1: TPopupActionBar;
    Exit1: TMenuItem;
    comboboxStyle: TComboBox;
    Panel5: TPanel;
    Label1: TLabel;
    Button1: TButton;
    ColorBox1: TColorBox;
    Panel4: TPanel;
    Panel6: TPanel;
    Button2: TButton;
    ColorBox2: TColorBox;
    FDGUIxWaitCursor1: TFDGUIxWaitCursor;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure NDBGrid1AfterAutoSet(Sender: TObject);
    procedure comboboxStyleChange(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure NDBGrid1MouseMove(Sender: TObject; Shift: TShiftState; X, Y: Integer);
    procedure DBGrid1MouseMove(Sender: TObject; Shift: TShiftState; X, Y: Integer);
  private
    FDmDatabase: TdmDatabase;
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

procedure TForm1.FormCreate(Sender: TObject);
var
  StyleNames: TStringList;
  StyleName: string;
begin
  FDmDatabase := TdmDatabase.Create(nil);
  DataSource1.DataSet := FDmDatabase.FDQuery1;

  ColorBox1.Selected := NDBGrid1.TitleFont.Color;
  ColorBox2.Selected := NDBGrid1.FixedColor;

  StyleNames := TStringList.Create;
  try
    for StyleName in TStyleManager.StyleNames do
      StyleNames.Append(StyleName);
    comboboxStyle.Items.Assign(StyleNames);
  finally
    StyleNames.Free;
  end;
  comboboxStyle.ItemIndex :=
    comboboxStyle.Items.IndexOf(TStyleManager.ActiveStyle.Name);

  FDmDatabase.FDQuery1.Open;
end;

procedure TForm1.FormDestroy(Sender: TObject);
begin
  FDmDatabase.Free;
end;

procedure TForm1.comboboxStyleChange(Sender: TObject);
begin
  TStyleManager.SetStyle(comboboxStyle.Text);
end;

procedure TForm1.NDBGrid1AfterAutoSet(Sender: TObject);
var
  Grid: TNDBGrid;
  OrderByClause: string;
  DataSet: TFDQuery;
  CmdText: TStringList;
  Re: TRegEx;
  Idx: Integer;
begin
  Grid := Sender as TNDBGrid;
  OrderByClause := Grid.GetOrderByClause;
  if not (Grid.DataSource.DataSet is TFDQuery) then
    Exit;

  DataSet := TFDQuery(Grid.DataSource.DataSet);
  CmdText := TStringList.Create;
  try
    CmdText.Assign(DataSet.SQL);

    DataSet.Close;
    try
      OutputDebugString(PChar(Grid.OrderByPrefix));

      Re := TRegEx.Create('^\s*' + Grid.OrderByPrefix + '.*$');
      for Idx := 0 to CmdText.Count - 1 do
        if Re.IsMatch(CmdText[Idx]) then begin
          CmdText[Idx] := OrderByClause;
          DataSet.SQL.Assign(CmdText);
          Exit;
        end;

      CmdText.Append(OrderByClause);
      DataSet.SQL.Assign(CmdText);
    finally
      DataSet.Open;
    end;
  finally
    CmdText.Free;
  end;
end;

procedure TForm1.Button1Click(Sender: TObject);
begin
  NDBGrid1.TitleFont.Color := ColorBox1.Selected;
  DBGrid1.TitleFont.Color := ColorBox1.Selected;
end;

procedure TForm1.Button2Click(Sender: TObject);
begin
  NDBGrid1.FixedColor := ColorBox2.Selected;
  DBGrid1.FixedColor := ColorBox2.Selected;
end;

procedure TForm1.NDBGrid1MouseMove(Sender: TObject; Shift: TShiftState;
  X, Y: Integer);
begin
  if (not NDBGrid1.Sizing(X, Y)) and (ssLeft in Shift) then
    NDBGrid1.BeginDrag(False);
end;

procedure TForm1.DBGrid1MouseMove(Sender: TObject; Shift: TShiftState;
  X, Y: Integer);
begin
end;

end.
